Add unit tests for utility functions

mimischi opened the pull request:
The PR adds the first unit tests for the utility functions found in `utils.py`.


## Todo

- ~~Add test for `utils.guess_ncores`~~

PR Checklist
------------
 - [x] CHANGELOG updated?

kain88-de commented on setup.cfg:
```diff
@@ -1,3 +1,13 @@
+
+[tool:pytest]
+testpaths = mdbenchmark/tests
+addopts = --cov=./
```
please remove coverage line. we don't want to do this everytime.

kain88-de commented on setup.cfg:
```diff
@@ -1,3 +1,13 @@
+test=pytest
+
+[tool:pytest]
+testpaths = mdbenchmark/tests
```
pytest can figure this out itself. I would remove this line.

kain88-de replied:
that way it's also more future resilient to changes in the file structure!

kain88-de commented on setup.cfg:
```diff
@@ -1,3 +1,13 @@
+addopts = --cov=./
+
+[coverage:run]
+omit = mdbenchmark/tests/*
```
we want to know if our tests are all run!

mimischi replied:
Isn't that the job of `pytest`? It will give you a list of tests that ran and those that were skipped/failed..?

kain88-de commented on setup.py:
```diff
@@ -28,8 +28,16 @@
+        'xdg<2',
     ],
     entry_points={'console_scripts': ['mdbenchmark=mdbenchmark.cli:cli']},
+    setup_requires=['pytest-runner'],
```
setup doesn't need pytest

mimischi replied:
I added that in case we use `python setup.py test` to run the tests, as I had some problems with setting up `pytest` initially.

kain88-de commented on setup.py:
```diff
@@ -28,8 +28,16 @@
     ],
     entry_points={'console_scripts': ['mdbenchmark=mdbenchmark.cli:cli']},
+    setup_requires=['pytest-runner'],
+    tests_require=['pytest', 'mock', 'pytest-cov'],
```
please remove coverage for the tests. This is only useful for CI runs and occasionally on a dev machine. Adding coverage usually makes tests slower to run.

kain88-de replied:
is the mock really needed?

mimischi replied:
We need `mock` to test the responses from system packages like `socket.gethostname`. This way we can test it on all machines. If we wouldn't do that, then we'd need to write tests specifically for Travis-CI and these would fail on any other machines.
See for example: https://github.com/bio-phys/MDBenchmark/pull/3/files#diff-4c190294c14d18b75f1b5a206c73944cR32

mimischi replied:
Oh wait. I just realized that we're anyway using the `monkeypatch` fixture from `pytest`. It's possible that we can actually get rid of the `mock` dependency then.

kain88-de commented on mdbenchmark/tests/test_utils.py:
```diff
@@ -0,0 +1,96 @@
+
+    slope_intercept = calc_slope_intercept(x1, y1, x2, y2)
+
+    np.testing.assert_array_almost_equal(slope_intercept,
```
use top level import of assert function. Also `assert_almost_equal` is more strict then the `assert_array_*` functions. Don't ask me why numpy devs did this.

kain88-de commented on .travis.yml:
```diff
@@ -15,6 +15,5 @@ install:
-
 script:
-  - true
+  - python setup.py test
```
if you add tests with coverage information please enable coveralls or codecov as well

mimischi replied:
So remove the coverage completely or keep it? Also you'd need to enable one of the integrations for the `bio-phys` organization, I can't.

kain88-de replied:
sure? Codecov doesn't seem to need any special permissions. You have to change the travis config. Have a look at `MDAnalysis/pytng`

mimischi replied:
Just checked, yes. It's exactly like with Travis-CI. You grant the access for the whole organization once and then all repositories can use it afterwards.

kain88-de replied:
should work now

kain88-de commented on .travis.yml:
```diff
@@ -14,7 +14,7 @@ branches:
 
 script:
-  - true
+  - pytest --cov=./
```
`pytest --cov mdbenchmark -v --pep8 mdbenchmark`

this means you also need to install `pytest-pep8`

mimischi commented:
Should I squash some commits before merging?

kain88-de commented on .travis.yml:
```diff
@@ -14,7 +14,7 @@ branches:
 
 script:
-  - true
+  - pytest --cov mdbenchmark -v --pep8 mdbenchmark
```
there still needs to be a 

```yaml
install:
   - pip install codecov

after_success:
    - codecov
```

kain88-de commented:
Yeah squashing would be nice.

kain88-de commented:
once we submit data to codecov you can also add a coverage token to the readme.

kain88-de commented:
oh and update the CHANGELOG

kain88-de commented:
@mimischi you can squash and merge when the tests pass

mimischi commented:
Just updated the CHANGELOG, bumped the version to `3.1.0` and squashed some commits. Looks like Codecov is having some hickups, so we may need to wait for it to finally finish the report.

I didn't write any tests for `utils.guess_ncores`, as I don't see how that should be that in a sensible way..?

I'm fine merging this PR.

kain88-de commented:
@mimischi no need to bump the version. We can actually accumulate thing between releases.

kain88-de commented:
I removed the version bump!

I will see on sunday how to publish a release on pypi and add some docs and tools to do that. I also want to experiment with `towncrier` to generate the changelog before we release.

mimischi commented:
Sounds good. I'll later add another test for the missing method in `utils.py` and then merge, when Codecov finishes.

kain88-de commented:
you can merge without codecov. That's OK as long as travis passes
